packages/teams-js/src/public/meeting.ts

```diff
@@ -334,6 +334,47 @@ export namespace meeting {
      * @returns A promise with the updated microphone state
      */
     micMuteStateChangedCallback: (micState: MicState) => Promise<MicState>;
+    /**
+     * Callback for the host to tell the app to change its speaker selection
+     */
+    audioDeviceSelectionChangedCallback?: (selectedDevices: AudioDeviceSelection) => void;
+  }
+
+  /**
+   * Interface for AudioDeviceSelection from host selection
+   *
+   * @hidden
+   * Hide from docs.
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface AudioDeviceSelection {
+    speaker?: AudioDeviceInfo;
+    microphone?: AudioDeviceInfo;
+    error?: SdkError;
+  }
```

Comment:
Continuing our conversation from the other PR...

I totally understand that a developer can certainly cast use `as any` and completely remove any type safety. We don't need to design or optimize for developers who are intentionally eschewing the safety provided by TypeScript. However, we should design our APIs so that developers who are trying to leverage TypeScript can do so in a way whereby it's hard or impossible to use the API/data structures in the wrong way. The design of the data should match the semantics: if it's not possible to have both an error and audio data, then the data structure should model that reality, enforced by the type system.

Here is an example of how leveraging the type system makes it impossible for a developer to use the data in the wrong way (assuming they aren't trying to intentionally subvert the type system):

```
function isSdkError(potentialError: unknown): potentialError is SdkError {
  return (potentialError as SdkError).errorCode !== undefined;
}

export interface AudioDeviceSelection {
  speaker?: AudioDeviceInfo;
  microphone?: AudioDeviceInfo;
}

interface RequestAppAudioHandlingParams {
...
  audioDeviceSelectionChangedCallback?: (resultOrError: AudioDeviceSelection | SdkError) => void;
}

function devCallback(selectedDevicesOrError: AudioDeviceSelection | SdkError): void {
  if (isSdkError(selectedDevicesOrError)) {
    // handle error
  }

  if (selectedDevicesOrError.microphone) {
    // the above line will not compile if an SdkError was returned; TypeScript recognizes that to access the microphone property, the type has to first be checked to ensure it is not SdkError
  }
}
```

Naturally there are other designs that also achieve the same goal (e.g., using a union on the object passed to the callback, rather than passing an AudioDeviceSelection | SdkError) -- but the common theme across the designs is that it enables the TypeScript compiler to prevent misuse of the data and the data models the semantics of what is and is not possible.

Reply:
Updated.